pkg/git/libgit2/managed/transport.go

```diff
@@ -0,0 +1,100 @@
+package managed
+
+import (
+	"crypto/md5"
+	"crypto/sha1"
+	"crypto/sha256"
+	"fmt"
+	"hash"
+	"net"
+
+	pkgkh "github.com/fluxcd/pkg/ssh/knownhosts"
+	git2go "github.com/libgit2/git2go/v33"
+	"golang.org/x/crypto/ssh/knownhosts"
+)
+
+// knownHostCallback returns a CertificateCheckCallback that verifies
+// the key of Git server against the given host and known_hosts for
+// git.SSH Transports.
+func KnownHostsCallback(host string, knownHosts []byte) git2go.CertificateCheckCallback {
+	return func(cert *git2go.Certificate, valid bool, hostname string) error {
+		kh, err := pkgkh.ParseKnownHosts(string(knownHosts))
+		if err != nil {
+			return fmt.Errorf("failed to parse known_hosts: %w", err)
+		}
+
+		// First, attempt to split the configured host and port to validate
+		// the port-less hostname given to the callback.
+		hostWithoutPort, _, err := net.SplitHostPort(host)
+		if err != nil {
+			// SplitHostPort returns an error if the host is missing
+			// a port, assume the host has no port.
+			hostWithoutPort = host
+		}
+
+		// Different versions of libgit handle this differently.
+		// This fixes the case in which ports may be sent back.
+		hostnameWithoutPort, _, err := net.SplitHostPort(hostname)
+		if err != nil {
+			hostnameWithoutPort = hostname
+		}
+
+		if hostnameWithoutPort != hostWithoutPort {
+			return fmt.Errorf("host mismatch: %q %q", hostWithoutPort, hostnameWithoutPort)
+		}
+
+		// We are now certain that the configured host and the hostname
+		// given to the callback match. Use the configured host (that
+		// includes the port), and normalize it, so we can check if there
+		// is an entry for the hostname _and_ port.
+		h := knownhosts.Normalize(host)
+		var fingerprint []byte
+		var hasher hash.Hash
+		switch {
+		case cert.Hostkey.Kind&git2go.HostkeySHA256 > 0:
+			fingerprint = cert.Hostkey.HashSHA256[:]
+			hasher = sha256.New()
+		case cert.Hostkey.Kind&git2go.HostkeySHA1 > 0:
+			fingerprint = cert.Hostkey.HashSHA1[:]
+			hasher = sha1.New()
+		case cert.Hostkey.Kind&git2go.HostkeyMD5 > 0:
+			fingerprint = cert.Hostkey.HashMD5[:]
+			hasher = md5.New()
```

Comment:
We are only keeping this just for unmanaged transport, so it would be good to add a comment to that effect.

Reply:
As per the other comment, we could also remove this.